sqlglot/dialects/tsql.py

```diff
@@ -719,9 +719,19 @@ def _parse_user_defined_function(
             ):
                 return this
 
-            expressions = self._parse_csv(self._parse_function_parameter)
+            expressions = None
+            if self._curr.token_type != TokenType.WITH:
+                expressions = self._parse_csv(self._parse_function_parameter)
+
             return self.expression(exp.UserDefinedFunction, this=this, expressions=expressions)
 
+        def _parse_function_parameter(self) -> t.Optional[exp.Expression]:
+            id_var = self._parse_id_var()
+            kind = self._parse_types(schema=True)
+            if not kind:
+                return id_var
+            return self.expression(exp.ColumnDef, this=id_var, kind=kind)
```

Comment:
What triggered this override? It looks pretty similar to the base implementation.

Reply:
This method in the Parser class is:
```
def _parse_function_parameter(self) -> t.Optional[exp.Expression]:
        return self._parse_column_def(self._parse_id_var())
```
Inside of the `_parse_column_def` method, `_parse_column_constraint` is called. This method parses the tokens as column constraints if they match the `CONSTRAINT_PARSERS` of the Parser class. One of these constraint parsers is for the `WITH` keyword, so if I leave the original implementation, the parser tries to parse procedure options as column constraints and eventually raises an error. That is why I gave TSQL its own `_parse_function_parameter` implementation.

I could replace the implementation of `_parse_function_parameter` in TSQL with an implementation of `_parse_column_constraint` that checks that the current token is not `WITH`, like so: 

```
def _parse_column_constraint(self) -> t.Optional[exp.Expression]:
    if self._match(TokenType.CONSTRAINT):
        this = self._parse_id_var()
    else:
        this = None

    # Check that WITH does not match and then keep going
    if not self._match(TokenType.WITH, advance=False) and self._match_texts(self.CONSTRAINT_PARSERS):
        return self.expression(
            exp.ColumnConstraint,
            this=this,
            kind=self.CONSTRAINT_PARSERS[self._prev.text.upper()](self),
        )

    return this
```

Reply:
I'm not sure my override of `_parse_function_parameter` in the TSQL class respects the spec even if all the tests pass. That is why I'm also thinking of overriding `_parse_column_constraint` instead, since that would preserve the functionality of the original implementation of `_parse_function_parameter` in the Parser class.